feat: held item cooldown overlay + focused mob health overlay

phantamanta44 opened the pull request:
## Features

This PR adds two new overlays, both of which are disabled by default:

* Held item cooldown overlay: displays an indicator for the cooldown of the held item (e.g. from using your main attack).
  * Also includes a feature that override vanilla crosshair rendering with our own; otherwise, the crosshair is off-center. This is toggleable as a separate feature.
  * Also includes a normal function `held_item_cooldown` for accessing the held item cooldown.

* Focused mob health bar overlay: displays a custom replacement for the focused mob health bar, which is usually displayed as a boss bar
  * Useful for when players want to move that information elsewhere on the screen.
  * Also includes normal functions for accessing the focused mob's name, health, and elemental properties (as an unparsed string for now).

[Here's a brief video demonstration of these overlays.](https://www.youtube.com/watch?v=rIhDmUHa0VE)

## Implementation Notes

* The properties of the focused mob are captured by `DamageModel`, which is responsible for parsing the focused mob health bar as a boss bar.
  * Some high-health mobs display their health as an abbreviated number, so there's a new utility function for parsing such numbers.
    * Since we support suffixes up to the trillions, mob health is now represented as a `long`. **This is technically a breaking change!** However, nothing in Wynntils currently uses these private values, so it's unlikely that anything was actually broken.
* The focused mob's elemental properties are currently retrieved as the raw string extracted from the boss bar. In this form, it's usable for displaying as a text overlay, but it'll need to be parsed into something more structured if players want to use it for template logic. This is left as future work.

phantamanta44 committed:
feat: additional overlay features

* Adds the held item cooldown overlay
  * Includes a fix for the vanilla crosshair positioning issue
  * Includes a normal function for accessing the held item cooldown
* Adds the focused mob health bar overlay
  * Includes normal functions for accessing the focused mob's name, health, and elemental properties

BREAKING CHANGE: mob health is now represented as a `long` in the `DamageModel`

Rafii2198 commented:
>  In this form, it's usable for displaying as a text overlay, but it'll need to be parsed into something more structured if players want to use it for template logic

If you mean that the new `focused_mob_health()` function would return something like `12345678900` and want to have it abbrivated then the `format()` function can do that. In this case `format(focused_mob_health)` would return `12.3M` given the number from the example.
If I understand your function correctly ofc.

Rafii2198 commented:
Also both types of functions in FunctionMenager are not sorted alphabetically. Same for features.

phantamanta44 commented:
> > In this form, it's usable for displaying as a text overlay, but it'll need to be parsed into something more structured if players want to use it for template logic
> 
> If you mean that the new `focused_mob_health()` function would return something like `12345678900` and want to have it abbrivated then the `format()` function can do that. In this case `format(focused_mob_health)` would return `12.3M` given the number from the example. If I understand your function correctly ofc.

that last bullet point is referring to the mob's elemental properties, i.e. its weaknesses, resistances, and damage types



> Also both types of functions in FunctionMenager are not sorted alphabetically. Same for features.

aren't the features in alphabetic order? and i noticed that not all the functions in `FunctionManager` are alphabetically ordered, so i assumed that it was supposed to be grouped by feature

Rafii2198 commented:
My bad on the sorting of features, I probably missread it, as for functions I believe they are grouped by type and sorted alphabetically kind of like features.

I also read again that last bullet point and my conclusion is that I should leave such reviews to devs especially at 3am... My bad, sorry.

kristofbolyai commented on common/src/main/java/com/wynntils/features/overlays/FixCrosshairPositionFeature.java:
```diff
@@ -0,0 +1,54 @@
+import net.neoforged.bus.api.EventPriority;
+import net.neoforged.bus.api.SubscribeEvent;
+
+@ConfigCategory(Category.OVERLAYS)
```
Isn't this more of an UTILITY than an OVERLAY?

kristofbolyai replied:
However, this begs the question, whether we should fix this, or just make users use separate mods, if they care to

phantamanta44 replied:
wasn't sure about the category; it's technically a "vanilla overlay" of sorts. at any rate, i included this change because the off-center vanilla reticle made the centered item cooldown indicator look terrible

kristofbolyai replied:
I think that's why it should move to the utilities category, because the OVERLAY category is strictly reserved for our custom overlays.

kristofbolyai commented on common/src/main/java/com/wynntils/functions/InventoryFunctions.java:
```diff
@@ -310,4 +313,25 @@ protected List<String> getAliases() {
+            int remaining = cooldown.endTime - cooldowns.tickCount;
+            if (remaining <= 0) return CappedValue.EMPTY;
+
+            return new CappedValue(remaining, cooldown.endTime - cooldown.startTime);
```
Extract this to the appropriate model, functions should essentially be "public getters" of core components data.

kristofbolyai commented on common/src/main/java/com/wynntils/models/damage/DamageModel.java:
```diff
@@ -14,26 +14,30 @@
-    private static final Pattern DAMAGE_BAR_PATTERN =
-            Pattern.compile("^§[ac](.*) - §c(\\d+)§4❤(?:§r -( (§.(.+))(Dam|Weak|Def))+)?$");
+    private static final Pattern DAMAGE_BAR_PATTERN = Pattern.compile(
+            "^\\s*§[ace](.*) - §c(\\d+(?:\\.\\d+)?[kKmM]?)§4❤(?:§r - ( ?(§.(.+))(Dam|Weak|Def))+)?\\s*$");
```
Does Wynn do the "kKmM" now? We used to have to add that client side.

phantamanta44 replied:
i found it on a few mobs in SE with the abbreviation feature disabled

kristofbolyai commented on common/src/main/java/com/wynntils/utils/StringUtils.java:
```diff
@@ -19,6 +19,7 @@
 
 public final class StringUtils {
     private static final String[] SUFFIXES = {"", "k", "m", "b", "t"}; // kilo, million, billion, trillion (short scale)
+    private static final long[] SUFFIX_MULTIPLIERS = {-1L, 1_000L, 1_000_000L, 1_000_000_000L, 1_000_000_000_000L};
```
Why not 1L? I see that it is not used, but multiplying with 1L wouldn't mess up any math, even if you miss-handled something.

kristofbolyai commented on common/src/main/java/com/wynntils/models/damage/DamageModel.java:
```diff
@@ -76,22 +96,31 @@ public double getAverageAreaDamagePerSecond(int seconds) {
+            // Wynncraft updates the focused mob health bar by repeatedly destroying and recreating the boss bar. We
+            // grab the progress here, when the bar's recreation event is received, to avoid the brief delay before
+            // the first progress update event is received, during which the state would otherwise be inconsistent
+            onUpdateProgress(event.getProgress());
```
This sounds like something that should be handled on a `BarHandler` (if that's how it's called) level. I'll leave this note here, and take a look, or alternatively, you can too, if you want to get this PR in faster.

phantamanta44 replied:
i initially considered just dealing with it in `BossBarHandler`, but i was hesitant to make what is technically a breaking behavioural change, especially considering the comment here
https://github.com/Wynntils/Wynntils/blob/1a3e676a8ad51c3e4bd1ec4e160d966d2ba24d4c/common/src/main/java/com/wynntils/models/abilities/bossbars/BloodPoolBar.java#L30-L32
which seems to imply that not updating progress immediately is the expected behaviour

kristofbolyai replied:
Okay, sounds sane enough.

magicus replied:
The comment in BloodPoolBar is just odd? I think bolyai has a point in that this would be better handled in the BossBarHandler. Can you try doing it there anyway and see if it breaks BloodPool bars? Otherwise I think this is basically a "bug fix" of BossBarHandler.

phantamanta44 replied:
i'll give it a shot, but admittedly i'm not much of a shaman player lol

magicus replied:
You can always ask for help testing on Discord if you need.

kristofbolyai reviewed (commented):
Just as an overall comment, didn't the actual "attack cooldown" feature used to work with Vanilla? Or was the timing wrong? Maybe it could be patched and fixed, and that would mean much less work for us.

phantamanta44 commented:
> Just as an overall comment, didn't the actual "attack cooldown" feature used to work with Vanilla? Or was the timing wrong? Maybe it could be patched and fixed, and that would mean much less work for us.

there is the vanilla swing speed indicator for weapon swings, but that's a separate subsystem from item cooldowns, and as far as i know, it's never been used by Wynncraft. in principle, it might be possible to reuse it, but it would require either an ugly mixin hack or for us to just re-render it entirely, which kinda defeats the point of reusing it in the first place

(also i just like the way mine looks more lol)

kristofbolyai commented:
Sounds reasonable, I originally didn't check how it looks, but now I agree it looks much better than the vanilla one.

kristofbolyai commented on common/src/main/java/com/wynntils/overlays/gamebars/FocusedMobHealthBarOverlay.java:
```diff
@@ -0,0 +1,78 @@
+        String healthString = Managers.Feature.getFeatureInstance(AbbreviateMobHealthFeature.class)
+                        .isEnabled()
+                ? StringUtils.integerToShortString(Models.Damage.getFocusedMobHealth())
+                : Long.toString(Models.Damage.getFocusedMobHealth());
```
Should you use `Long.toString`? IIRC it adds thousand separators and in some locales it looks really weird.

phantamanta44 replied:
it shouldn't; it's the normal toString map for `long`s

phantamanta44 committed:
fix: fix issues with "more overlays" implementation

* Tracks focused mob state more correctly in `DamageModel`
* Moves the crosshair position fix to the "utilities" category
* Factors item cooldown out of the inventory function and into the character stats model
* Changes the fallback value for `StringUtils::SUFFIX_MULTIPLIERS` to 1

phantamanta44 committed:
ci: spotless formatting

kristofbolyai commented on common/src/main/java/com/wynntils/features/utilities/FixCrosshairPositionFeature.java:
```diff
@@ -0,0 +1,54 @@
+
+@ConfigCategory(Category.UTILITIES)
+public class FixCrosshairPositionFeature extends Feature {
+    private static boolean shouldOverrideCrosshair(Minecraft mc) {
```
It's a bit weird to pass in MC, no? You really could just access it statically.

phantamanta44 committed:
chore: move `Minecraft` access into `FixCrosshairPositionFeature::shouldOverrideCrosshair`

phantamanta44 committed:
fix: allow any colour mob name in `DamageModel` health bar parser

phantamanta44 commented:
ran a raid yesterday and realized that raid bosses have dark red and bold names, so i've just allowed any colour in the health bar regexp

magicus commented on common/src/main/java/com/wynntils/overlays/HeldItemCooldownOverlay.java:
```diff
@@ -0,0 +1,76 @@
+        renderOverlay(poseStack, bufferSource, 1F);
+    }
+
+    private void renderOverlay(PoseStack poseStack, MultiBufferSource bufferSource, float cdPercent) {
```
Can I ask you to spell out `coolDownPercent`?

magicus commented on common/src/main/java/com/wynntils/overlays/HeldItemCooldownOverlay.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Override
+    protected void onConfigUpdate(Config<?> config) {}
```
Do we still need to implement `onConfigUpdate` for overlays? I thought we fixed that.

phantamanta44 replied:
it does still appear to be defined `abstract` in `Overlay`, yes

magicus replied:
*sigh* Ok.

kristofbolyai replied:
I can fix that.

magicus commented:
Overall I think this looks great! Both in terms of the functionality provided, and the quality of the source code. 

Thanks for contributing to Wynntils.

phantamanta44 committed:
chore: `cdPercent` -> `cooldownPercent`

phantamanta44 committed:
feat!: emit first boss bar progress update on bar creation in `BossBarHandler`

While this is technically a breaking change, it shouldn't break behaviour in any meaningful way; if anything, it should remove the possibility of ephemeral inconsistent states between the creation of a boss bar and the arrival of the first progress update

kristofbolyai commented on common/src/main/java/com/wynntils/models/damage/DamageModel.java:
```diff
@@ -60,6 +71,37 @@ public void onLabelIdentified(LabelIdentifiedEvent event) {
+
+    public String getFocusedMobElementals() {
+        checkFocusedMobValidity();
+        return focusedMobElementals; // TODO: Parse this into specific elements and expose as functions
```
We usually don't do same-line TODOS, but really just a nit.

```suggestion
        // TODO: Parse this into specific elements and expose as functions
        return focusedMobElementals;
```

kristofbolyai committed:
fix: todo line

kristofbolyai commented:
This also looks really good, thanks! I'll go ahead and test the blood pool bar, then merge this.

kristofbolyai commented:
Blood pool does work! I'll merge this now (without doing a release just yet.)
